File: sdks/swift/Sources/PuxbaySDK/Puxbay.swift
import Foundation

/// Main Puxbay API client with URLSession and retry logic
public class Puxbay {
    private let apiKey: String
    private let baseURL: URL
    private let session: URLSession
    private let maxRetries: Int
    
    /// Initialize Puxbay client
    /// - Parameters:
    ///   - apiKey: API key (must start with 'pb_')
    ///   - baseURL: Base URL for API (default: https://api.puxbay.com/api/v1)
    ///   - timeout: Request timeout in seconds (default: 30)
    ///   - maxRetries: Maximum retry attempts (default: 3)
    public init(
        apiKey: String,
        baseURL: String = "https://api.puxbay.com/api/v1",
        timeout: TimeInterval = 30,
        maxRetries: Int = 3
    ) throws {
        guard apiKey.hasPrefix("pb_") else {
            throw PuxbayError.invalidAPIKey
        }
        
        guard let url = URL(string: baseURL) else {
            throw PuxbayError.invalidBaseURL
        }
        
        self.apiKey = apiKey
        self.baseURL = url
        self.maxRetries = maxRetries
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.httpAdditionalHeaders = [
            "X-API-Key": apiKey,
            "Content-Type": "application/json",
            "User-Agent": "puxbay-swift/1.0.0",
            "Accept-Encoding": "gzip, deflate"
        ]
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        config.urlCache = nil
        
        self.session = URLSession(configuration: config)
    }
    
    // Resource accessors
    public lazy var products = ProductsResource(client: self)
    public lazy var orders = OrdersResource(client: self)
    public lazy var customers = CustomersResource(client: self)
    public lazy var categories = CategoriesResource(client: self)
    public lazy var suppliers = SuppliersResource(client: self)
    public lazy var giftCards = GiftCardsResource(client: self)
    public lazy var branches = BranchesResource(client: self)
    public lazy var staff = StaffResource(client: self)
    public lazy var webhooks = WebhooksResource(client: self)
    public lazy var inventory = InventoryResource(client: self)
    public lazy var reports = ReportsResource(client: self)
    public lazy var purchaseOrders = PurchaseOrdersResource(client: self)
    public lazy var stockTransfers = StockTransfersResource(client: self)
    public lazy var stocktakes = StocktakesResource(client: self)
    public lazy var cashDrawers = CashDrawersResource(client: self)
    public lazy var expenses = ExpensesResource(client: self)
    public lazy var notifications = NotificationsResource(client: self)
    public lazy var returns = ReturnsResource(client: self)
    
    /// Make HTTP request with retry logic
    func request<T: Decodable>(
        method: String,
        endpoint: String,
        body: Encodable? = nil
    ) async throws -> T {
        var lastError: Error?
        
        for attempt in 0...maxRetries {
            do {
                let url = baseURL.appendingPathComponent(endpoint)
                var request = URLRequest(url: url)
                request.httpMethod = method
                
                if let body = body {
                    let encoder = JSONEncoder()
                    encoder.keyEncodingStrategy = .convertToSnakeCase
                    request.httpBody = try encoder.encode(body)
                }
                
                let (data, response) = try await session.data(for: request)
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw PuxbayError.invalidResponse
                }
                
                // Retry on 429 or 5xx
                if httpResponse.statusCode == 429 || httpResponse.statusCode >= 500 {
                    if attempt < maxRetries {
                        let delay = pow(2.0, Double(attempt))
                        try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                        continue
                    }
                }
                
                guard (200...299).contains(httpResponse.statusCode) else {
                    try handleErrorResponse(statusCode: httpResponse.statusCode, data: data)
                }
                
                if T.self == EmptyResponse.self {
                    return EmptyResponse() as! T
                }
                
                let decoder = JSONDecoder()
                decoder.keyDecodingStrategy = .convertFromSnakeCase
                return try decoder.decode(T.self, from: data)
            } catch {
                lastError = error
                if attempt < maxRetries {
                    let delay = pow(2.0, Double(attempt))
                    try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                }
            }
        }
        
        throw lastError ?? PuxbayError.unknownError
    }
    
    private func handleErrorResponse(statusCode: Int, data: Data) throws {
        let message: String
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        
        if let errorResponse = try? decoder.decode(ErrorResponse.self, from: data) {
            message = errorResponse.detail ?? errorResponse.message ?? "Unknown error"
        } else {
            message = "Unknown error"
        }
        
        switch statusCode {
        case 401:
            throw PuxbayError.authenticationError(message)
        case 429:
            throw PuxbayError.rateLimitError(message)
        case 400:
            throw PuxbayError.validationError(message)
        case 404:
            throw PuxbayError.notFoundError(message)
        case 500...:
            throw PuxbayError.serverError(message)
        default:
            throw PuxbayError.apiError(message, statusCode)
        }
    }
}

// MARK: - Error Types

public enum PuxbayError: Error {
    case invalidAPIKey
    case invalidBaseURL
    case invalidResponse
    case authenticationError(String)
    case rateLimitError(String)
    case validationError(String)
    case notFoundError(String)
    case serverError(String)
    case apiError(String, Int)
    case unknownError
}

File: sdks/php/src/Puxbay.php
<?php

namespace Puxbay;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use Puxbay\Exceptions\PuxbayException;
use Puxbay\Exceptions\AuthenticationException;
use Puxbay\Exceptions\RateLimitException;
use Puxbay\Exceptions\ValidationException;
use Puxbay\Exceptions\NotFoundException;
use Puxbay\Exceptions\ServerException;
use Puxbay\Resources\BranchesResource;
use Puxbay\Resources\CashDrawersResource;
use Puxbay\Resources\CategoriesResource;
use Puxbay\Resources\CustomersResource;
use Puxbay\Resources\ExpensesResource;
use Puxbay\Resources\GiftCardsResource;
use Puxbay\Resources\InventoryResource;
use Puxbay\Resources\NotificationsResource;
use Puxbay\Resources\OrdersResource;
use Puxbay\Resources\ProductsResource;
use Puxbay\Resources\PurchaseOrdersResource;
use Puxbay\Resources\ReportsResource;
use Puxbay\Resources\ReturnsResource;
use Puxbay\Resources\StaffResource;
use Puxbay\Resources\StocktakesResource;
use Puxbay\Resources\StockTransfersResource;
use Puxbay\Resources\SuppliersResource;
use Puxbay\Resources\WebhooksResource;

/**
 * Main Puxbay API client with connection pooling and retry logic.
 *
 * @package Puxbay
 */
class Puxbay
{
    private const DEFAULT_BASE_URL = 'https://api.puxbay.com/api/v1';
    private const SDK_VERSION = '1.0.0';

    private GuzzleClient $client;
    private string $apiKey;
    private string $baseUrl;
    private int $maxRetries;

    // Resource clients
    public readonly ProductsResource $products;
    public readonly OrdersResource $orders;
    public readonly CustomersResource $customers;
    public readonly CategoriesResource $categories;
    public readonly SuppliersResource $suppliers;
    public readonly GiftCardsResource $giftCards;
    public readonly BranchesResource $branches;
    public readonly StaffResource $staff;
    public readonly WebhooksResource $webhooks;
    public readonly InventoryResource $inventory;
    public readonly ReportsResource $reports;
    public readonly PurchaseOrdersResource $purchaseOrders;
    public readonly StockTransfersResource $stockTransfers;
    public readonly StocktakesResource $stocktakes;
    public readonly CashDrawersResource $cashDrawers;
    public readonly ExpensesResource $expenses;
    public readonly NotificationsResource $notifications;
    public readonly ReturnsResource $returns;

    /**
     * Create a new Puxbay client.
     *
     * @param string $apiKey API key (must start with 'pb_')
     * @param array $config Configuration options
     * @throws \InvalidArgumentException
     */
    public function __construct(string $apiKey, array $config = [])
    {
        if (!str_starts_with($apiKey, 'pb_')) {
            throw new \InvalidArgumentException("Invalid API key format. Must start with 'pb_'");
        }

        $this->apiKey = $apiKey;
        $this->baseUrl = $config['base_url'] ?? self::DEFAULT_BASE_URL;
        $this->maxRetries = $config['max_retries'] ?? 3;
        $timeout = $config['timeout'] ?? 30;

        // Create retry middleware
        $handlerStack = HandlerStack::create();
        $handlerStack->push(Middleware::retry(
            function ($retries, $request, $response, $exception) {
                if ($retries >= $this->maxRetries) {
                    return false;
                }

                // Retry on 429 or 5xx errors
                if ($response && ($response->getStatusCode() === 429 || $response->getStatusCode() >= 500)) {
                    return true;
                }

                // Retry on connection errors
                if ($exception instanceof GuzzleException) {
                    return true;
                }

                return false;
            },
            function ($retries) {
                // Exponential backoff: 1s, 2s, 4s, 8s...
                return (int) (1000 * pow(2, $retries));
            }
        ));

        // Create Guzzle client
        $this->client = new GuzzleClient([
            'base_uri' => $this->baseUrl,
            'timeout' => $timeout,
            'handler' => $handlerStack,
            'headers' => [
                'X-API-Key' => $this->apiKey,
                'Content-Type' => 'application/json',
                'User-Agent' => 'puxbay-php/' . self::SDK_VERSION,
                'Accept-Encoding' => 'gzip, deflate',
            ],
        ]);

        // Initialize resource clients
        $this->products = new ProductsResource($this);
        $this->orders = new OrdersResource($this);
        $this->customers = new CustomersResource($this);
        $this->categories = new CategoriesResource($this);
        $this->suppliers = new SuppliersResource($this);
        $this->giftCards = new GiftCardsResource($this);
        $this->branches = new BranchesResource($this);
        $this->staff = new StaffResource($this);
        $this->webhooks = new WebhooksResource($this);
        $this->inventory = new InventoryResource($this);
        $this->reports = new ReportsResource($this);
        $this->purchaseOrders = new PurchaseOrdersResource($this);
        $this->stockTransfers = new StockTransfersResource($this);
        $this->stocktakes = new StocktakesResource($this);
        $this->cashDrawers = new CashDrawersResource($this);
        $this->expenses = new ExpensesResource($this);
        $this->notifications = new NotificationsResource($this);
        $this->returns = new ReturnsResource($this);
    }

    /**
     * Make an HTTP request to the API.
     *
     * @param string $method HTTP method
     * @param string $endpoint API endpoint
     * @param array|null $body Request body
     * @return array Response data
     * @throws PuxbayException
     */
    public function request(string $method, string $endpoint, ?array $body = null): array
    {
        try {
            $options = [];
            if ($body !== null) {
                $options['json'] = $body;
            }

            $response = $this->client->request($method, $endpoint, $options);
            $content = (string) $response->getBody();

            return $content ? json_decode($content, true) : [];
        } catch (GuzzleException $e) {
            if ($e->hasResponse()) {
                $response = $e->getResponse();
                $statusCode = $response->getStatusCode();
                $body = (string) $response->getBody();

                $this->handleErrorResponse($statusCode, $body);
            }

            throw new PuxbayException('Network error: ' . $e->getMessage(), 0);
        }
    }

    /**
     * Handle error responses.
     *
     * @param int $statusCode HTTP status code
     * @param string $body Response body
     * @throws PuxbayException
     */
    private function handleErrorResponse(int $statusCode, string $body): void
    {
        $message = 'Unknown error';
        $data = json_decode($body, true);

        if ($data && isset($data['detail'])) {
            $message = $data['detail'];
        } elseif ($data && isset($data['message'])) {
            $message = $data['message'];
        }

        switch ($statusCode) {
            case 401:
                throw new AuthenticationException($message, $statusCode);
            case 429:
                throw new RateLimitException($message, $statusCode);
            case 400:
                throw new ValidationException($message, $statusCode);
            case 404:
                throw new NotFoundException($message, $statusCode);
            default:
                if ($statusCode >= 500) {
                    throw new ServerException($message, $statusCode);
                }
                throw new PuxbayException($message, $statusCode);
        }
    }
}
